staging/src/k8s.io/apimachinery/pkg/util/validation/field/testing/match.go

```diff
@@ -0,0 +1,203 @@
+		for _, e := range remaining {
+			t.Errorf("unmatched error:\n%s", Match().Exactly().Render(e))
+		}
+	}
+}
+
+// Match returns a new Matcher.
+func Match() *Matcher {
+	return &Matcher{}
+}
+
+// Matcher is a helper for comparing field.Error objects.
+type Matcher struct {
+	matchType                bool
+	matchField               bool
+	matchValue               bool
+	matchOrigin              bool
+	matchDetail              func(want, got string) bool
+	requireOriginWhenInvalid bool
+}
+
+// Matches returns true if the two field.Error objects match according to the
+// configured criteria.
+func (m *Matcher) Matches(want, got *field.Error) bool {
+	if m.matchType && want.Type != got.Type {
+		return false
+	}
+	if m.matchField && want.Field != got.Field {
+		return false
+	}
+	if m.matchValue && !reflect.DeepEqual(want.BadValue, got.BadValue) {
+		return false
+	}
+	if m.matchOrigin {
+		if want.Origin != got.Origin {
+			return false
+		}
+		if m.requireOriginWhenInvalid && want.Type == field.ErrorTypeInvalid {
+			if want.Origin == "" || got.Origin == "" {
+				return false
+			}
+		}
+	}
+	if m.matchDetail != nil && !m.matchDetail(want.Detail, got.Detail) {
+		return false
+	}
+	return true
+}
+
+// Render returns a string representation of the specified Error object,
+// according to the criteria configured in the Matcher.
+func (m *Matcher) Render(e *field.Error) string {
+	buf := strings.Builder{}
+
+	comma := func() {
+		if buf.Len() > 0 {
+			buf.WriteString(", ")
+		}
+	}
+
+	if m.matchType {
+		comma()
+		buf.WriteString(fmt.Sprintf("Type=%q", e.Type))
+	}
+	if m.matchField {
+		comma()
+		buf.WriteString(fmt.Sprintf("Field=%q", e.Field))
+	}
+	if m.matchValue {
+		comma()
+		buf.WriteString(fmt.Sprintf("Value=%v", e.BadValue))
+	}
+	if m.matchOrigin || m.requireOriginWhenInvalid && e.Type == field.ErrorTypeInvalid {
+		comma()
+		buf.WriteString(fmt.Sprintf("Origin=%q", e.Origin))
+	}
+	if m.matchDetail != nil {
+		comma()
+		buf.WriteString(fmt.Sprintf("Detail=%q", e.Detail))
+	}
+	return "{" + buf.String() + "}"
+}
+
+// Exactly configures the matcher to match all fields exactly.
+func (m *Matcher) Exactly() *Matcher {
+	return m.ByType().ByField().ByValue().ByOrigin().ByDetailExact()
+}
+
+// Exactly configures the matcher to match errors by type.
+func (m *Matcher) ByType() *Matcher {
+	m.matchType = true
+	return m
+}
+
+// ByField configures the matcher to match errors by field path.
+func (m *Matcher) ByField() *Matcher {
+	m.matchField = true
+	return m
+}
+
+// ByValue configures the matcher to match errors by the errant value.
+func (m *Matcher) ByValue() *Matcher {
+	m.matchValue = true
+	return m
+}
+
+// ByOrigin configures the matcher to match errors by the origin.
+func (m *Matcher) ByOrigin() *Matcher {
+	m.matchOrigin = true
+	return m
+}
+
+// RequireOriginWhenInvalid configures the matcher to require the Origin field
+// to be set when the Type is Invalid.
+func (m *Matcher) RequireOriginWhenInvalid() *Matcher {
```

Comment:
`Matches` will only honor  `requireOriginWhenInvalid` when `matchOrigin` is true?  Should we document this or otherwise protect users from accidently leaving `matchOrigin` false when enabling `requireOriginWhenInvalid` ?

Reply:
Good callout.  Will document